// @walletconnect/browser-utils/dist/esm/browser.js
import * as windowMetadata from "@walletconnect/window-metadata";
import * as windowGetters from "@walletconnect/window-getters";
import { detect } from "detect-browser";
export function detectEnv(userAgent) {
  return detect(userAgent);
}
export function detectOS() {
  var env = detectEnv();
  return env && env.os ? env.os : undefined;
}
export function isAndroid() {
  var os = detectOS();
  return os ? os.toLowerCase().includes("android") : false;
}
export function isIOS() {
  var os = detectOS();
  return os ? os.toLowerCase().includes("ios") || os.toLowerCase().includes("mac") && navigator.maxTouchPoints > 1 : false;
}
export function isMobile() {
  var os = detectOS();
  return os ? isAndroid() || isIOS() : false;
}
export function isNode() {
  var env = detectEnv();
  var result = env && env.name ? env.name.toLowerCase() === "node" : false;
  return result;
}
export function isBrowser() {
  var result = !isNode() && !!getNavigator();
  return result;
}
export var getFromWindow = windowGetters.getFromWindow;
export var getFromWindowOrThrow = windowGetters.getFromWindowOrThrow;
export var getDocumentOrThrow = windowGetters.getDocumentOrThrow;
export var getDocument = windowGetters.getDocument;
export var getNavigatorOrThrow = windowGetters.getNavigatorOrThrow;
export var getNavigator = windowGetters.getNavigator;
export var getLocationOrThrow = windowGetters.getLocationOrThrow;
export var getLocation = windowGetters.getLocation;
export var getCryptoOrThrow = windowGetters.getCryptoOrThrow;
export var getCrypto = windowGetters.getCrypto;
export var getLocalStorageOrThrow = windowGetters.getLocalStorageOrThrow;
export var getLocalStorage = windowGetters.getLocalStorage;
export function getClientMeta() {
  return windowMetadata.getWindowMetadata();
}
