// @walletconnect/browser-utils/dist/esm/mobile.js
import { setLocal } from "./local";
export var mobileLinkChoiceKey = "WALLETCONNECT_DEEPLINK_CHOICE";
export function formatIOSMobile(uri, entry) {
  var encodedUri = encodeURIComponent(uri);
  return entry.universalLink ? "".concat(entry.universalLink, "/wc?uri=").concat(encodedUri) : entry.deepLink ? "".concat(entry.deepLink).concat(entry.deepLink.endsWith(":") ? "//" : "/", "wc?uri=").concat(encodedUri) : "";
}
export function saveMobileLinkInfo(data) {
  var focusUri = data.href.split("?")[0];
  setLocal(mobileLinkChoiceKey, Object.assign(Object.assign({}, data), {
    href: focusUri
  }));
}
export function getMobileRegistryEntry(registry, name) {
  return registry.filter(function (entry) {
    return entry.name.toLowerCase().includes(name.toLowerCase());
  })[0];
}
export function getMobileLinkRegistry(registry, whitelist) {
  var links = registry;

  if (whitelist) {
    links = whitelist.map(function (name) {
      return getMobileRegistryEntry(registry, name);
    }).filter(Boolean);
  }

  return links;
}
