// @walletconnect/browser-utils/dist/esm/registry.js
var API_URL = "https://registry.walletconnect.com";
export function getWalletRegistryUrl() {
  return API_URL + "/api/v2/wallets";
}
export function getDappRegistryUrl() {
  return API_URL + "/api/v2/dapps";
}
export function formatMobileRegistryEntry(entry) {
  var platform = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "mobile";

  var _a;

  return {
    name: entry.name || "",
    shortName: entry.metadata.shortName || "",
    color: entry.metadata.colors.primary || "",
    logo: (_a = entry.image_url.sm) !== null && _a !== void 0 ? _a : "",
    universalLink: entry[platform].universal || "",
    deepLink: entry[platform].native || ""
  };
}
export function formatMobileRegistry(registry) {
  var platform = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "mobile";
  return Object.values(registry).filter(function (entry) {
    return !!entry[platform].universal || !!entry[platform].native;
  }).map(function (entry) {
    return formatMobileRegistryEntry(entry, platform);
  });
}
